### ISSUE TITLE
Dependency resolution should not unnecessarily lock the target of a project dependency

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle-native#904]
Dependency resolution locks the target project of a project dependency in order to calculate project metadata, or to resolve the artifacts and files of a configuration, if not already done. However, when this happens at task execution time, a long running task in the target project  will block the worker thread.

We should take advantage of the fact that a task has declared that it will (indirectly) need these pieces of metadata in order to execute to prepare the metadata in some way before starting the task.

We should also verify that locking does in fact happen _only_ when the particular value is not already available.

**Comment by adammurdoch** [comment]
Some clean up in https://github.com/gradle/gradle/pull/7493.

I'd like to do a few more tidy-ups before closing this.

**Comment by adammurdoch** [comment]
Some more clean up in https://github.com/gradle/gradle/pull/7552. This will do for now, I think.
